### PR TITLE
[clang][analyzer] Fix #embed crash

### LINKED ISSUE #107724: clang-tidy crashes when it encounters #embed in C code
I ran into the following crash trying to run clang-tidy on some code using C23 `#embed` to include a JSON schema file into a `const char []`. The shortened output from clang-tidy (the [full-output](https://github.com/user-attachments/files/16920084/clang-tidy19-embed-crash.txt) is attached just in case):
```
LLVM ERROR: Support for EmbedExpr is not implemented.
PLEASE submit a bug report to https://github.com/llvm/llvm-project/issues/ and include the crash backtrace.
Stack dump:
0.	Program arguments: /usr/local/llvm19/bin/clang-tidy main.c -- -I/usr/local/include -fblocks
1.	<eof> parser at end of file
2.	While analyzing stack:
	#0 Calling get_schema at line 99
	#1 Calling main
3.	/home/crest/friendly-s6/main.c:1074:3: Error evaluating statement
4.	/home/crest/friendly-s6/main.c:1074:3: Error evaluating statement
```

Here's the trimmed down code to the function that triggers the error (with line numbers to match up against the crash report).
```
  1070 static inline ucl_object_t *_Nonnull
  1071 get_schema(void)
  1072 {
  1073         const unsigned char schema_json[] = {
  1074                 #embed "schema.json"
  1075         };
```

### PR DESCRIPTION
**Pull request by nicovank**
Fix #107724.

### CODE REVIEW

**Comment by nicovank** [comment]
PS: Should this be cherry-picked into 19?

**Review by NagyDonat — COMMENTED** [review]
LGTM, there is no reason to crash on `#embed`.

I think it would be nice to have a few testcases that show the behavior of the analyzer around `#embed`:
- Can we produce bug reports if there is an (unrelated) `#embed` expression on the execution path? Or do we abort the analysis there?
- If the analysis is not aborted, then how do we represent the contents of a memory region that's initialized by `#embed`? Is it `UnknownVal`? Or `UndefinedVal` (that may be problematic)?

File: clang/test/Analysis/embed.c [hunk 1]
```diff
@@ -0,0 +1,9 @@
+// RUN: %clang_analyze_cc1 -std=c23 -analyzer-checker=core,debug.ExprInspection -verify %s
+
+// expected-no-diagnostics
+
+int main() {
+    const unsigned char SelfBytes[] = {
+        #embed "embed.c"
+    };
+}
```
Comment (steakhal) [inline clang/test/Analysis/embed.c:9]:
```suggestion
// RUN: %clang_analyze_cc1 -std=c23 -analyzer-checker=core,debug.ExprInspection -verify %s

void clang_analyzer_dump_ptr(const unsigned char *ptr);
void clang_analyzer_dump(unsigned char val);

int main() {
    const unsigned char SelfBytes[] = {
        #embed "embed.c"
    };
    clang_analyzer_dump_ptr(SelfBytes); // expected-warning {{&Element{SelfBytes,0 S64b,unsigned char}}}
    clang_analyzer_dump(SelfBytes[0]); // expected-warning {{Unknown}} FIXME: This should be the `/` character.
}
```

**Review by steakhal — APPROVED** [review]
LGTM. I knew about this issue, thanks for the fix!
There is more for a proper fix but it's better than crashing.

**Comment by steakhal** [comment]
> PS: Should this be cherry-picked into 19?

Requested the backport in https://github.com/llvm/llvm-project/pull/107841

**Pull request merged**